django/db/models/functions/text.py

```diff
@@ -188,6 +190,133 @@ def as_sqlite(self, compiler, connection, **extra_context):
+    def _resolve_flags(self, flags, mapping, default=''):
+        value = default + flags.value
+
+        # Remove duplicate flags preserving the last occurrence of each to
+        # ensures that the last flag is preferred if any are contradictory.
+        value = ''.join(reversed(dict.fromkeys(reversed(value))))
+
+        # Resolve case-sensitivity flags preferring the last flag specified.
+        if {'c', 'i'}.issubset(value):
+            remove = 'i' if value.index('c') > value.index('i') else 'c'
+            value = value.replace(remove, '')
+
+        # Map flags from normalized values to backend-specific values:
+        flags.value = value.translate(str.maketrans(mapping))
+
+    def _resolve_extra(self, extra, flags):
+        extra.append(Value(1))  # position
+        extra.append(Value(0 if 'g' in flags.value else 1))  # occurrence
+        if isinstance(self, RegexpStrIndex):
+            extra.append(Value(0))  # return_opt
+        flags.value = flags.value.replace('g', '')
+        extra.append(flags)
+
+    def _inline_flags(self, pattern, flags, mariadb=False):
+        if flags.value and pattern.value:
+            # Force flags inline in the pattern instead of as an argument.
+            # Also convert case-sensitive flag to separate (?-i) for MariaDB.
+            if mariadb and 'c' in flags.value:
+                pattern.value = '(?-i)' + pattern.value
+                flags.value = flags.value.replace('c', '')
+            flags.value = flags.value.replace('g', '')
+            pattern.value = ('(?%s)' % flags.value) + pattern.value
+
+    def as_mysql(self, compiler, connection, **extra_context):
+        clone = self.copy()
+        expression, pattern, *extra, flags = clone.get_source_expressions()
+        # Force case-sensitive matching by default for consistency.
+        mapping = {} if connection.mysql_is_mariadb else {'s': 'n'}
+        self._resolve_flags(flags, mapping, default='c')
+        if not connection.mysql_is_mariadb:
+            self._resolve_extra(extra, flags)
+        else:
+            self._inline_flags(pattern, flags, mariadb=True)
+        clone.set_source_expressions([expression, pattern, *extra])
+        return clone.as_sql(compiler, connection, **extra_context)
+
+    def as_oracle(self, compiler, connection, **extra_context):
+        clone = self.copy()
+        expression, pattern, *extra, flags = clone.get_source_expressions()
+        self._resolve_flags(flags, {'s': 'n'})
+        self._resolve_extra(extra, flags)
+        clone.set_source_expressions([expression, pattern, *extra])
+        return clone.as_sql(compiler, connection, **extra_context)
+
+    def as_postgresql(self, compiler, connection, **extra_context):
+        clone = self.copy()
+        expression, pattern, *extra, flags = clone.get_source_expressions()
+        # Force partial newline-sensitive matching by default for consistency.
+        self._resolve_flags(flags, {'m': 'n'}, default='p')
```

Comment:
Can we move the set of default flags to the `DatabaseOperations` hook, e.g. `regexp_function_flags()` with an empty set in the default implementation? This should make `as_*` methods unnecessary, unless on PostgreSQL and SQLite.

Reply:
I _think_ I get what you're getting at with this, but will have to have a play around with it.

I'm not convinced we can save much (or anything) as there are some oddities, e.g. MariaDB uses `_inline_flags()` but it is also needed in one case for PostgreSQL and `_resolve_extra()` is only required for MySQL and Oracle. I don't see how `as_mysql()` and `as_oracle()` can be eliminated as you seem to suggest.